### PR TITLE
Add language: Mojo

### PR DESCRIPTION
**Pull request by lattner**
Add a new programming language to Linguist, named Mojo, for file extensions `.mojo` and `.🔥`. You can learn more about Mojo at https://docs.modular.com/mojo/.

## Description

Mojo is a new programming language developed by me ([Chris Lattner](https://en.wikipedia.org/wiki/Chris_Lattner)) and my amazing team at [Modular](http://modular.com/).

Because it's a brand new language, it does not meet linguist's requirements for broad use on GitHub. The team and I thought that we'd send this anyway, to see if an exception could be made, based on the success and widespread use on GitHub of some of my other programming languages (LLVM IR, MLIR, Swift). If that's not possible, no worries, and thanks for your time!

## Checklist:

- [x] **I am adding a new language.**
  - [ ] ~The extension of the new language is used in hundreds of repositories on GitHub.com.~ As explained above, this is a new language without extensive usage yet.
    - Search results for each extension:
      
      -  https://github.com/search?q=path%3A*.mojo&type=code
      -  [https://github.com/search?q=path%3A*.🔥&type=code](https://github.com/search?q=path%3A*.%F0%9F%94%A5&type=code)
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - `Matmul.mojo` is adapted from https://docs.modular.com/mojo/notebooks/Matmul.html
      - `Bool.🔥` is adapted from https://docs.modular.com/mojo/notebooks/BoolMLIR.html
    - Sample license(s):
  - [x] I have included a syntax highlighting grammar: https://github.com/modularml/mojo-syntax
      
  - [x] I have added a color
    - Hex value: `#ff4c1f`
    - Rationale:  This color invokes the 🔥 emoji that is used for the language's branding, and as a supported file extension.
  - [ ] ~I have updated the heuristics to distinguish my language from others using the same extension.~ No other language in this repository uses the `.mojo` or `.🔥` file extensions, so I did not add a heuristic.

### CODE REVIEW

**Comment by lildude** [comment]
> The team and I thought that we'd send this anyway, to see if an exception could be made, based on the success and widespread use on GitHub of some of my other programming languages (LLVM IR, MLIR, Swift).

Unfortunately, we can't as it sets a precedent  we don't want to fight each and every time someone develops "the next greatest thing, promise on my hamster's life". Given the popularity you've mentioned, and the hype around AI, it shouldn't take long to reach the levels we require though.

**Comment by lattner** [comment]
Makes sense, thank you for the consideration!

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -4193,6 +4193,15 @@ Module Management System:
   tm_scope: none
   ace_mode: text
   language_id: 235
+Mojo:
+  type: programming
+  color: "#ff4c1f"
+  extensions:
+  - ".mojo"
+  - ".\U0001F525"
+  ace_mode: text
```
Comment (Alhadis) [inline lib/linguist/languages.yml:4202]:
```suggestion
  ace_mode: python
  codemirror_mode: python
  codemirror_mime_type: text/x-python
```

Mojo is similar enough to Python that updates are pulled from [`MagicPython.tmBundle`](https://github.com/MagicStack/MagicPython/blob/master/grammars/MagicPython.tmLanguage), according to [comments](https://github.com/modularml/mojo-syntax/blob/81cf73ef2d4a1b149d58b4e7fab2f1acf4edde2d/syntaxes/mojo.syntax.json#L2-L6) in its grammar file. Ergo, we might as well repurpose the Python modes of CodeMirror and ACE (as the alternative here would be no syntax highlighting in those editors at all).

**Review by Alhadis — CHANGES_REQUESTED** [review]
An emoji file extension. I've seen everything now…

**Comment by lattner** [comment]
> An emoji file extension. I've seen everything now…

Can't stop the innovation! :)

File: lib/linguist/languages.yml (hunk 1 above)
Comment (modocache) [reply lib/linguist/languages.yml]: Done, thanks for your help! (We've also refreshed the samples included to use the latest Mojo syntax 😄)

**Comment by modocache** [comment]
Hey all! First off thanks for your reviews since we first submitted this pull request. Originally Chris wrote:

> Because it's a brand new language, it does not meet linguist's requirements for broad use on GitHub.

But I believe we may now meet those requirements. The contribution guidelines [here](https://github.com/github-linguist/linguist/blob/master/CONTRIBUTING.md#adding-a-language) say to look for 200 unique `:user/:repo` repositories, but that due to GitHub search being in flux, temporarily the requirement documented [here](https://github.com/github-linguist/linguist/issues/5756) is "at least 2000 files per extension indexed in the last year (the number you see at the top of the search results)." In a search for files named `*.mojo` [here](https://github.com/search?q=path%3A*.mojo&type=code), I see 1.1k results -- and there a few `.🔥` results as well.

Also, not sure if it's worth mentioning, but as of [this patch](https://github.com/vim/vim/commit/0ce2c594d0704f27a16d2c13fce85d596cc91489), Vim now recognizes the `.mojo` and `.🔥` extensions.

So, would it be possible to take another look at this and its "Pending Popularity" tag? Thanks in advance! 🙏  @lildude @Alhadis

**Comment by lildude** [comment]
No change. It's still doesn't look popular enough, especially when you consider that `mojo-lang` owns most of them so [excluding them](https://github.com/search?q=path%3A*.mojo+NOT+is%3Afork+NOT+user%3Amojo-lang&type=code) drops things quite considerably, And most definitely not the emoji extension. You might want to consider removing that extension from this PR as it will only delay things. Users can always implement an override until such time as it's popular which for addition in the future.

**Comment by lildude** [comment]
And further [refinement](https://github.com/search?q=path%3A*.mojo+NOT+is%3Afork+NOT+user%3Amojo-lang+NOT+user%3Amodularml+language%3AXML&type=code&l=XML) shows a lot of those remaining files are XML, which whilst not popular enough for inclusion right now, would need to be explicitly added to XML in this PR too to ensure they remain correctly identified as XML (the extension takes precedence so this PR as it stands would cause all those files to be identified as Mojo and incorrect syntax highlighting will silly).

**Comment by modocache** [comment]
Gotcha, thanks for the quick & thorough response! I'll use those searches to keep tabs on things, and in the meantime update this PR to handle XML with a `.mojo` extension.

> And most definitely not the emoji extension. You might want to consider removing that extension from this PR as it will only delay things.

I see, so I should consider the popularity of the `.mojo` extension as separate from the popularity of the `.🔥` extension. That makes sense, thanks!

**Comment by modocache** [comment]
Thanks for your help! I pushed an update here with heuristics for Mojo vs. XML, and removed `.🔥`.

I'll keep an eye on [the search link you shared](https://github.com/search?q=path%3A*.mojo+NOT+is%3Afork+NOT+user%3Amojo-lang+NOT+language%3AXML&type=code) and ping this thread when we cross the popularity threshold. Right now I see 1.1k public Mojo files that are not XML 😃

**Comment by guidorice** [comment]
I was just looking at the search link, and could not find my mojo repository in the results. Wondering how the indexing works, and whether it's an accurate count? It's https://github.com/guidorice/geo-features

**Comment by djkelleher** [comment]
This Google Sheet is a great way to view Mojo popularity: https://docs.google.com/spreadsheets/d/1mmS5xwRrtBIZubdEIrUpsGzZxX2mcqcc74tBjDfo6Lo/edit?usp=sharing (updated daily)
A quick look shows:
- there are currently about 310 unique (non-fork) public Mojo repos.
- the current rate of new Mojo file creation is about 3k a year. (1588 files/6.5ish months since launch)(12month/year) = 2932 files/year
- sum of forks for all unique repos is over 2.3k

**Comment by tairov** [comment]
@lildude I think we're good to go

image

**Comment by lildude** [comment]
> Hah, thank you to all of our community members pinging this -- it does indeed look like we're above the 2k source file threshold! Reviewers, please take a look at your earliest convenience, thanks!

Not quite. Three users are having an undue influence on the figures with one repo accounting for over 1000 of the results. [Excluding them and forks](https://github.com/search?q=path%3A*.mojo+NOT+IS%3Afork+NOT+user%3Asarvex++NOT+user%3Amojo-lang+NOT+language%3AXML+NOT+user%3AIvo-Balbaert&type=code) brings things down quite a bit.

**Comment by modocache** [comment]
Ah OK, thanks for the clarification, sorry if I missed that. Thanks!

**Comment by djkelleher** [comment]
I'll try to make this my last annoying comment :slightly_smiling_face:

I think the count shown in the [sheet](https://docs.google.com/spreadsheets/d/1mmS5xwRrtBIZubdEIrUpsGzZxX2mcqcc74tBjDfo6Lo/edit?usp=sharing) is more accurate than that query. There are more non-fork Mojo repos that show up when you leave off the `fork:false` parameter, and also more that show up using different queries and API endpoints, as well as a few that seem to be not indexed at all.

The way to get the most accurate count seemed to be to try a bunch of queries (via HTTP API), use the `fork` boolean in the JSON responses to filter out forks, then check the file content/extension.

Everything in the sheet has been gone through by multiple people and the file count as of today is 4585 with one outlier of 2579. Excluding outlier: 4585-2579=2006 files

P.S. you can scope queries to `created` after Mojo launch in the beginning of May 2023

**Comment by iamtimdavis** [comment]
@lildude - Just wanted to check if we need to do anything else here? Want to make sure we have everything you need. Thanks so much again

**Comment by lildude** [comment]
Almost... I can't update the PR to merge in master as maintainers haven't been granted write perms on this PR and can't approve CI to run for some reason. 

Usage looks good though.

File: lib/linguist/languages.yml (hunk 1 above)
Comment (Alhadis) [reply lib/linguist/languages.yml]: CodeMirror's [`python` mode](https://codemirror.net/5/mode/python/index.html) does a decent enough job of highlighting Mojo source, but it hasn't been included in the PR as I initially suggested.

~~~diff
 Mojo:
   type: programming
   color: "#ff4c1f"
   extensions:
   - ".mojo"
   ace_mode: python
   tm_scope: source.mojo
   language_id: 1045019587
+  codemirror_mode: python
+  codemirror_mime_type: text/x-python
~~~

@modocache or @lattner, could one of you please action this remaining change?

**Comment by modocache** [comment]
> Almost... I can't update the PR to merge in master as maintainers haven't been granted write perms on this PR and can't approve CI to run for some reason.

Amazing! Sorry, small parse error on my end for this sentence -- do I need to, or can I, update write permissions on this PR? Or are you saying that for some reason write permissions don't seem to have been granted by some automated process, and it's not something I can fix?

File: lib/linguist/languages.yml (hunk 1 above)
Comment (modocache) [reply lib/linguist/languages.yml]: Yes, can do right now, thanks! Sorry, missed this comment when I commented on the main thread earlier. Thanks!

Comment (modocache) [reply lib/linguist/languages.yml]: Done -- sorry again for missing this!

**Comment by lildude** [comment]
> > Almost... I can't update the PR to merge in master as maintainers haven't been granted write perms on this PR and can't approve CI to run for some reason.
> 
> Amazing! Sorry, small parse error on my end for this sentence -- do I need to, or can I, update write permissions on this PR? Or are you saying that for some reason write permissions don't seem to have been granted by some automated process, and it's not something I can fix?

It's something you do when creating the PR and can be updated on current PRs. See the [docs here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).

**Comment by lildude** [comment]
NM. You can't because your fork is owned by an org. It's only an option for personally owned forks.

**Review by lildude — APPROVED** [review]

**Review by Alhadis — APPROVED** [review]

File: lib/linguist/languages.yml (hunk 1 above)
Comment (Alhadis) [reply lib/linguist/languages.yml]: All good! Now we're ready to rock. 😉

**Comment by lattner** [comment]
Thank you @lildude !

**Pull request merged**